### ISSUE TITLE
Copy module failure in 2.20

### ISSUE THREAD

**Issue by andy-zhangtao** [ansible/ansible#15721]
##### ISSUE TYPE
- Bug Report
##### ANSIBLE VERSION

```
ansible-playbook 2.2.0 (devel 1fc44e4103) last updated 2016/05/02 07:18:40 (GMT +000)
  lib/ansible/modules/core: (detached HEAD b6ad3b6773) last updated 2016/05/02 07:18:47 (GMT +000)
  lib/ansible/modules/extras: (detached HEAD 1846de2809) last updated 2016/05/02 07:18:49 (GMT +000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/opt/ansible/ansible/library']
```
##### CONFIGURATION

[defaults]
forks          = 500
gathering = explicit
host_key_checking = False
log_path = /var/log/ansible.log
no_log = False

[privilege_escalation]
become=True
become_method=sudo
become_user=root
become_ask_pass=False

[ssh_connection]
ssh_args =
scp_if_ssh = True
##### OS / ENVIRONMENT

ansible/centos7-ansible:latest
##### SUMMARY

The copy module failure in remote machines.
##### STEPS TO REPRODUCE

The content of playbook.yml

---
- hosts: coreos
  tasks:
  - name: copy shell file
     copy: src=/tmp/delete.sh dest=/tmp/delete2.sh
    The content of Inventory:
    [coreos]
    10.11.3.115 ansible_ssh_private_key_file=/tmp/xxxx.pem ansible_ssh_user=core ansible_python_interpreter=/opt/bin/python

```
ansible-playbook -i Inventory playbook.yml
```
##### EXPECTED RESULTS

delete.sh will be copied to remote machine.
##### ACTUAL RESULTS

```
fatal: [10.11.3.115]: FAILED! => {"failed": true, "msg": "Failed to get information on remote file (/tmp/delete2.sh): MODULE FAILURE"}
```

**Comment by alikins** [comment]
(This comment is just a copy of the playbook in a markdown code formatting for easier reading).

```
  ---
  - hosts: coreos
    tasks:
       - name: copy shell file
         copy: src=/tmp/delete.sh dest=/tmp/delete2.sh
```

**Comment by alikins** [comment]
> - hosts: coreos

If the host is coreos, has python been set up on the remote system?

The "MODULE FAILURE" means the copy module itself failed. It looks like coreos doesn't have python installed by default (based on https://coreos.com/blog/managing-coreos-with-ansible/ ) so that could be causing the error.

Do other modules work?

**Comment by andy-zhangtao** [comment]
I use raw and synchronize, these modules works fine. I use ansible2.1.0 test. It also works fine.

ansible/centos7-ansible:latest is the official image? Should I build it myself?

**Comment by andy-zhangtao** [comment]
I get the below error. It seems the python has some wrong configure in remote machines.

```
/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8) 
/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8) 
/etc/localtime is not a symlink, not updating container timezone. 
Traceback (most recent call last): 
  File \"/home/core/.ansible/tmp/ansible-tmp-1462413293.33-173671562381843/file\", line 114, in <module> 
    exitcode = invoke_module(module, zipped_mod, ZIPLOADER_PARAMS) 
  File \"/home/core/.ansible/tmp/ansible-tmp-1462413293.33-173671562381843/file\", line 28, in invoke_module 
    p = subprocess.Popen(['/opt/bin/python', module], env=os.environ, shell=False, stdout=subprocess.PIPE, stderr=subprocess.PIPE, stdin=subprocess.PIPE) 
  File \"/usr/lib/python2.7/subprocess.py\", line 710, in __init__ 
    errread, errwrite) 
  File \"/usr/lib/python2.7/subprocess.py\", line 1335, in _execute_child 
    raise child_exception 
OSError: [Errno 2] No such file or directory 
```

**Comment by andy-zhangtao** [comment]
@alikins  I use file module re-test. It also has this issue. 

```
- file: path=/tmp/test state=directory mode=0755
```

I get two version(2.1.0 and 2.2.0) of the file(the file module script in remote machines). 

2.1.0 create folder directly, however 2.2.0 create another python script(ansible_module_file.py in the /tmp folder), then use that script to create folder. 

When python use 

```
    p = subprocess.Popen(['/opt/bin/python', module], env=os.environ, shell=False, stdout=subprocess.PIPE, stderr=subprocess.PIPE, stdin=subprocess.PIPE)
module is the path of ansible_module_file.py.
```

Although module save the script path, but python still can not find ansible_module_file.py.

I compare two cases of environment. They have some difference.
2.1.0 :

```
[{'TERM': 'xterm', 'container': 'systemd-nspawn', 'LOGNAME': 'root', 'USER': 'root', 'HOME': '/root', 'PATH': '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'}]
```

2.2.0:

```
[{'TERM': 'xterm', 'container': 'systemd-nspawn', 'PYTHONPATH': '/tmp/ansible_O0XYgi/ansible_modlib.zip', 'LOGNAME': 'root', 'USER': 'root', 'HOME': '/root', 'PATH': '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'}]
```

Do you think is a bug of ansible 2.2.0 or wrong configure in coreos?

**Comment by andy-zhangtao** [comment]
I modify executor/module_common.py. change 

```
p = subprocess.Popen([%(interpreter)s, module], env=os.environ, shell=False, stdout=subprocess.PIPE, stderr=subprocess.PIPE, stdin=subp
rocess.PIPE)
```

to

```
p = subprocess.Popen([%(interpreter)s, module], env=os.environ, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, stdin=subp
rocess.PIPE)
```

It seems has some effect. Now the error change to :

```
{
    "changed": false,
    "failed": true,
    "module_stderr": "",
    "module_stdout": "/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)\r\n/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)\r\n/etc/localtime is not a symlink, not updating container timezone.\r\n/tmp/ansible_QtpjsQ/ansible_module_file.py: line 1: /opt/bin/python: not found\r\n",
    "msg": "MODULE FAILURE",
    "parsed": false
}
```

/opt/bin/python is my coreos python path.

**Comment by alikins** [comment]
The traceback in https://github.com/ansible/ansible/issues/15721#issuecomment-217058095 and the 
test in https://github.com/ansible/ansible/issues/15721#issuecomment-217101157 make me suspect this is a ziploader issue, though there are a lot of variables in play.

Just to clarify that I understand the scenario:

1) 'ansible-playbook' is being run from a centos7 system
2) the 'coreos' host target is a https://coreos.com/ system
3) that target host 'coreos' has python install at /opt/bin/python

**Comment by alikins** [comment]
```
/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
/etc/localtime is not a symlink, not updating container timezone.
/tmp/ansible_QtpjsQ/ansible_module_file.py: line 1: /opt/bin/python: not found
```

What is is the locale setup on the system where ansible-playbook is running? (ie, is LANG or LC_ALL set and what is it set to?)
The LC_ALL warning usually means the remote system doesn't have the forwarded locale installed (and I suspect for this case, the remote system has almost no locales installed.

`/tmp/ansible_QtpjsQ/ansible_module_file.py: line 1: /opt/bin/python: not found`

If I understand correctly, the  module delivered to the remote machine is being run with '/opt/bin/python', but it doesn't seem to be found.

Changing shell=True making it work makes me think there is some setup going on in the remote node via shell profiles (guesses... does coreos need to add or mount /opt/ from the login profile?). ie, with shell=true, is there there additional env setup that is skipped without a shell.

**Comment by alikins** [comment]
How is /opt/bin/python setup on the remote coreos node?

https://github.com/ianbytchek/docker-coreos-ansible-toolbox?

via coreos toolbox? https://coreos.com/os/docs/latest/install-debugging-tools.html

other?

**Comment by andy-zhangtao** [comment]
Yes, /opt/bin/python has setup via the toolbox on the coreos node. 
I don't know what's the use of shell=true, but it works in my side.

**Comment by milaw** [comment]
Have a same issue
Trying to access to coresos from centos 7

**Comment by milaw** [comment]
fatal: [127.0.0.1]: FAILED! => {"changed": false, "failed": true, "module_stderr": "", "module_stdout": "Traceback (most recent call last):\r\n  File \"/home/ansible/.ansible/tmp/ansible-tmp-1467118668.72-129864102443735/setup\", line 114, in <module>\r\n    exitcode = invoke_module(module, zipped_mod, ZIPLOADER_PARAMS)\r\n  File \"/home/ansible/.ansible/tmp/ansible-tmp-1467118668.72-129864102443735/setup\", line 28, in invoke_module\r\n    p = subprocess.Popen(['PATH=/usr/bin/python:$PATH python', module], env=os.environ, shell=False, stdout=subprocess.PIPE, stderr=subprocess.PIPE, stdin=subprocess.PIPE)\r\n  File \"/usr/local/lib/python2.7/subprocess.py\", line 710, in **init**\r\n    errread, errwrite)\r\n  File \"/usr/local/lib/python2.7/subprocess.py\", line 1335, in _execute_child\r\n    raise child_exception\r\nOSError: [Errno 2] No such file or directory\r\n", "msg": "MODULE FAILURE", "parsed": false}

**Comment by rwky** [comment]
It's due to this line https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/module_common.py#L124 and is beyond just the copy module, on CoreOS the `ansible_python_interpreter` variable is often set to include PATH i.e `PATH=/usr/bin/python:$PATH python` as per @milaw has above. This doesn't work in the `subprocess.Popen` function, python tries to use the entire `ansible_python_interpreter` variable as the python binary which of course doesn't work.

Work around for CoreOS users is to symlink (or directly copy) python to `/opt/bin/python`

**Comment by simfarm** [comment]
I've run into the same issue.

**Comment by ansibot** [comment]
@andy-zhangtao ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ghost** [comment]
This issue exists on a clean install of Ubuntu 16.04 server.

**Comment by ros0x5Ft** [comment]
Ubuntu 16.04.1 LTS (GNU/Linux 4.4.0-36-generic x86_64)

**Comment by ansibot** [comment]
@andy-zhangtao ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by tedder** [comment]
@andy-zhangtao why did you close this issue? It wasn't fixed. cc @alikins bc we should probably open this.

**Comment by andy-zhangtao** [comment]
I mistakenly thought it needed to be closed.

**Comment by tedder** [comment]
Okay. This happens in `setup` also, and probably elsewhere. @abadger, I think these two sections don't agree with each other in the ziploader: 

[section A](https://github.com/ansible/ansible/commit/452034564c32be415f1d2289339bf7bbbf23d273#diff-321e6923b14b571aad803d04b3308a35R205), [section B](https://github.com/ansible/ansible/commit/452034564c32be415f1d2289339bf7bbbf23d273#diff-321e6923b14b571aad803d04b3308a35L200)

By doing a bunch of hacking I've found that the tempdir contains the py and a zipped version of the ansible libs it needs to run. But they never get unzipped. In the code above, B sends along the path to the zipped libs, but it never gets unzipped.

This is a breaking problem on coreos with pypy, I'm not sure if it's a problem elsewhere.

**Comment by tedder** [comment]
This doesn't seem to be due to the `/opt/bin/python` thing. I symlinked it, but Ansible is respecting my python path, it appears:

    p = subprocess.Popen(['PATH=/home/core/bin:$PATH', '/home/core/bin/python', module],
     env=os.environ, shell=False, stdout=subprocess.PIPE, stderr=subprocess.PIPE, stdin=subprocess.PIPE)

**Comment by tedder** [comment]
Okay, got it. A couple of things.

First, `ansible_python_interpreter` can't have a PATH or anything in it. Those are environment vars, and Ansible can't handle that. It works without any env vars, though.

Next, the subprocess thing is (I think) because ANSIBALLZ makes the assumption that Python can import the ansible libs while they are in a zip. I didn't even know that was a thing, and I've had trouble googling for it. Anyhow, if I monkeypatch module_common.py to unzip the zip to the tempdir, it works.

So the next step is to fix module_common.py, I think? I mean, it needs a flag or a fallback for this error case, I'm guessing.

**Comment by abadger** [comment]
Every version of python from 2.4, through 2.7 can import from a zip.

**Comment by tedder** [comment]
I don't think pypy can, though.

**Comment by abadger** [comment]
Works with: pypy-5.4.0-1.fc25.x86_64

```
$ ansible -m ping -a 'data=werewrew' localhost -e 'ansible_python_interpreter=/usr/bin/pypy' -vvvv
Using /etc/ansible/ansible.cfg as config file
Loading callback plugin minimal of type stdout, v2.0 from /srv/ansible/abadger2/lib/ansible/plugins/callback/__init__.pyc
Using module file /srv/ansible/abadger2/lib/ansible/modules/core/system/ping.py
<localhost> ESTABLISH SSH CONNECTION FOR USER: None
<localhost> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/badger/.ansible/cp/ansible-ssh-%C localhost '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo ~/.ansible/tmp/ansible-tmp-1480379044.71-181609309022989 `" && echo ansible-tmp-1480379044.71-181609309022989="` echo ~/.ansible/tmp/ansible-tmp-1480379044.71-181609309022989 `" ) && sleep 0'"'"''
<localhost> PUT /var/tmp/tmpBUoPJX TO /home/badger/.ansible/tmp/ansible-tmp-1480379044.71-181609309022989/ping.py
<localhost> SSH: EXEC sftp -b - -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/badger/.ansible/cp/ansible-ssh-%C '[localhost]'
<localhost> ESTABLISH SSH CONNECTION FOR USER: None
<localhost> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/badger/.ansible/cp/ansible-ssh-%C localhost '/bin/sh -c '"'"'chmod u+x /home/badger/.ansible/tmp/ansible-tmp-1480379044.71-181609309022989/ /home/badger/.ansible/tmp/ansible-tmp-1480379044.71-181609309022989/ping.py && sleep 0'"'"''
<localhost> ESTABLISH SSH CONNECTION FOR USER: None
<localhost> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/badger/.ansible/cp/ansible-ssh-%C -tt localhost '/bin/sh -c '"'"'/usr/bin/pypy /home/badger/.ansible/tmp/ansible-tmp-1480379044.71-181609309022989/ping.py; rm -rf "/home/badger/.ansible/tmp/ansible-tmp-1480379044.71-181609309022989/" > /dev/null 2>&1 && sleep 0'"'"''
localhost | SUCCESS => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "data": "werewrew"
        }, 
        "module_name": "ping"
    }, 
    "ping": "werewrew"
}
```

**Comment by abadger** [comment]
@tedder You on IRC?  We can try to figure out if this is a specific version of pypy or something else (I'm abadger1999 on #ansible-devel)

**Comment by abadger** [comment]
I'm unable to reproduce this problem so far.  Here's what I did to test:

* Followed the download instructions for the stable channel of coreos here: https://coreos.com/os/docs/latest/booting-with-qemu.html
* Started the coreos vm: ./coreos_production_qemu.sh -a ~/.ssh/authorized_keys -- -nographic
* Ran ansible-galaxy install defunctzombie.coreos-bootstrap -p /srv/ansible/troubleshooting/roles to bootstrap the coreos host with python.
* Added this to my inventory:
```
[test]
coreos ansible_host=172.17.0.1 ansible_port=2222

[test:vars]
ansible_python_interpreter=/home/core/bin/python
ansible_user=core
```

made the following playbook:
```
---
- hosts: coreos
  tasks:
  - copy: src=copy.yml dest=/home/core/copy.yml
```

Ran ansible-playbook test.yml
```
root@fcb280e0d40c:~# ANSIBLE_SSH_PIPELINING=1 ansible-playbook test.yml -i inven -vvvvv
No config file found; using defaults
Loading callback plugin default of type stdout, v2.0 from /usr/local/lib/python2.7/dist-packages/ansible/plugins/callback/__init__.pyc

PLAYBOOK: test.yml *************************************************************
1 plays in test.yml

PLAY [coreos] ******************************************************************

TASK [setup] *******************************************************************
Using module file /usr/local/lib/python2.7/dist-packages/ansible/modules/core/system/setup.py
<172.17.0.1> ESTABLISH CONNECTION FOR USER: core on PORT 2222 TO 172.17.0.1
<172.17.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1480385617.89-273440187355394 `" && echo ansible-tmp-1480385617.89-273440187355394="` echo $HOME/.ansible/tmp/ansible-tmp-1480385617.89-273440187355394 `" ) && sleep 0'
<172.17.0.1> PUT /tmp/tmphcudnQ TO /home/core/.ansible/tmp/ansible-tmp-1480385617.89-273440187355394/setup.py
<172.17.0.1> EXEC /bin/sh -c 'chmod u+x /home/core/.ansible/tmp/ansible-tmp-1480385617.89-273440187355394/ /home/core/.ansible/tmp/ansible-tmp-1480385617.89-273440187355394/setup.py && sleep 0'
<172.17.0.1> EXEC /bin/sh -c '/home/core/bin/python /home/core/.ansible/tmp/ansible-tmp-1480385617.89-273440187355394/setup.py; rm -rf "/home/core/.ansible/tmp/ansible-tmp-1480385617.89-273440187355394/" > /dev/null 2>&1 && sleep 0'
ok: [coreos]

TASK [copy] ********************************************************************
task path: /root/test.yml:4
looking for "copy.yml" at "/root/files/copy.yml"
looking for "copy.yml" at "/root/copy.yml"
Using module file /usr/local/lib/python2.7/dist-packages/ansible/modules/core/files/stat.py
<172.17.0.1> ESTABLISH CONNECTION FOR USER: core on PORT 2222 TO 172.17.0.1
<172.17.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1480385618.62-194298722894194 `" && echo ansible-tmp-1480385618.62-194298722894194="` echo $HOME/.ansible/tmp/ansible-tmp-1480385618.62-194298722894194 `" ) && sleep 0'
<172.17.0.1> PUT /tmp/tmpCNiUY1 TO /home/core/.ansible/tmp/ansible-tmp-1480385618.62-194298722894194/stat.py
<172.17.0.1> EXEC /bin/sh -c 'chmod u+x /home/core/.ansible/tmp/ansible-tmp-1480385618.62-194298722894194/ /home/core/.ansible/tmp/ansible-tmp-1480385618.62-194298722894194/stat.py && sleep 0'
<172.17.0.1> EXEC /bin/sh -c '/home/core/bin/python /home/core/.ansible/tmp/ansible-tmp-1480385618.62-194298722894194/stat.py; rm -rf "/home/core/.ansible/tmp/ansible-tmp-1480385618.62-194298722894194/" > /dev/null 2>&1 && sleep 0'
<172.17.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1480385619.05-73054029258493 `" && echo ansible-tmp-1480385619.05-73054029258493="` echo $HOME/.ansible/tmp/ansible-tmp-1480385619.05-73054029258493 `" ) && sleep 0'
<172.17.0.1> PUT /root/copy.yml TO /home/core/.ansible/tmp/ansible-tmp-1480385619.05-73054029258493/source
<172.17.0.1> EXEC /bin/sh -c 'chmod u+x /home/core/.ansible/tmp/ansible-tmp-1480385619.05-73054029258493/ /home/core/.ansible/tmp/ansible-tmp-1480385619.05-73054029258493/source && sleep 0'
Using module file /usr/local/lib/python2.7/dist-packages/ansible/modules/core/files/copy.py
<172.17.0.1> PUT /tmp/tmpckUNhp TO /home/core/.ansible/tmp/ansible-tmp-1480385619.05-73054029258493/copy.py
<172.17.0.1> EXEC /bin/sh -c 'chmod u+x /home/core/.ansible/tmp/ansible-tmp-1480385619.05-73054029258493/ /home/core/.ansible/tmp/ansible-tmp-1480385619.05-73054029258493/copy.py && sleep 0'
<172.17.0.1> EXEC /bin/sh -c '/home/core/bin/python /home/core/.ansible/tmp/ansible-tmp-1480385619.05-73054029258493/copy.py; rm -rf "/home/core/.ansible/tmp/ansible-tmp-1480385619.05-73054029258493/" > /dev/null 2>&1 && sleep 0'
changed: [coreos] => {
    "changed": true, 
    "checksum": "4c1dc0f9392c03945192d3e3ef83ffdd1f574c75", 
    "dest": "/home/core/copy.yml", 
    "gid": 500, 
    "group": "core", 
    "invocation": {
        "module_args": {
            "backup": false, 
            "content": null, 
            "delimiter": null, 
            "dest": "/home/core/copy.yml", 
            "directory_mode": null, 
            "follow": false, 
            "force": true, 
            "group": null, 
            "mode": null, 
            "original_basename": "copy.yml", 
            "owner": null, 
            "regexp": null, 
            "remote_src": null, 
            "selevel": null, 
            "serole": null, 
            "setype": null, 
            "seuser": null, 
            "src": "/home/core/.ansible/tmp/ansible-tmp-1480385619.05-73054029258493/source", 
            "unsafe_writes": null, 
            "validate": null
        }
    }, 
    "md5sum": "7ac36bbca2a053f12736b180e9f727cb", 
    "mode": "0644", 
    "owner": "core", 
    "size": 6, 
    "src": "/home/core/.ansible/tmp/ansible-tmp-1480385619.05-73054029258493/source", 
    "state": "file", 
    "uid": 500
}

PLAY RECAP *********************************************************************
coreos                     : ok=2    changed=1    unreachable=0    failed=0  
```

Ansible version is:
```
root@fcb280e0d40c:~# ansible-playbook --version
ansible-playbook 2.2.0.0
  config file = 
  configured module search path = Default w/o overrides
```

**Comment by abadger** [comment]
I tried again with the coreos beta image and still wasn't able to reproduce this.

The only differences between what I've tried testing and what tedder has that we've identified so far:

* tedder has pypy-5.6.0.  I have pypy-5.1.0.  I'm not sure how to get the new pypy.
* I'm using the qemu image and tedder seems to be using the amazon image.

(also, for all those using environment variables in ansible_python_interpreter.... AFAIK, that's never worked.  If you need environment variables set, upload a script to the remote host which sets the environment variables and then set the script to be the ansible_python_interpreter.)

**Comment by sivel** [comment]
Copying my response from the mailing list here as well:

From the original report, it looks like this is due to trying to use toolbox to run python on coreos.  I believe this is due to having a shell script at /opt/bin/python that tries to execute /usr/bin/python inside of the toolbox container:

```
core@coreos ~ $ /opt/bin/python -c 'import sys; print sys.executable'
/usr/bin/python
```

This is what causes the issues where "No such file or directory" is reported.  The "python" path on the coreos machine is in a different location than the actual python binary in the toolbox container, so when `/opt/bin/python` is executed in the container, it fails.

In a simple test to see if this could be resolved, I linked /usr/bin/python to /opt/bin/python inside the toolbox container, and it resolves the issues:

```
$ toolbox --quiet /bin/sh
/bin/sh: can't access tty; job control turned off
/ # mkdir -p /opt/bin
/ # ln -s /usr/bin/python /opt/bin/
```

On the other hand, using the coreos instructions for installing pypy via https://github.com/defunctzombie/ansible-coreos-bootstrap work like a charm with ansible, with no complexities, as also confirmed by @abadger

**Comment by abadger** [comment]
@tedder confirmed on IRC that he was able to get ansible working with pypy on coreos via defunctzombie/ansible-coreos-bootstrap:

> [22:06:48] <tedder> test results: removed ansible.cfg, still fails. tried beta+prod, fails both. next: roll back some defunktzombie versions.
> [22:12:58] <tedder> that fixes it.

Thanks to all for the confirmation and walking through testing things alongside me.  I'll close this ticket now.  If other people are having issues that they think are bugs please open a new ticket for us to look at.  If you aren't sure if it's a bug or working as designed, please talk it over on IRC or the mailing list and then open a new ticket if applicable.  Thanks!

**Comment by tedder** [comment]
My case ended up being a missing `ansible_python_interpreter`, which isn't handled well in defunctzombie. It took a lot of troubleshooting to get there, though.

**Comment by NicholasPCole** [comment]
In case anyone else comes across this "module failure" error in the future and is scratching their head like I was, trying things like specifying `ansible_python_interpreter` to no success... make sure the host you’re connecting to has the `sudo` package installed. It wasn't by default on my Debian 9 box and led to that error when trying to `become`.

**Comment by gaozhidf** [comment]
the problem happens when the client is ubuntu16.04,  and install python 2.7 to fix the problem in the client.
